### DISCUSSION TITLE
Does intent-driven API mean the functions names need to sound like intent?

### DISCUSSION THREAD

**Discussion by derberg** [asyncapi/parser-api#4] [General]
While I love the idea of focusing on intent instead of AsyncAPI spec structure, I'm confused by the API because of function names. Do names need to sound like an intent? for me intent means that "I can do with API what I intend to do, do not have to learn the spec, just need to know the basic elements (channels, messages, etc) and do not need to know where are they in the specification, I mean where in the structure".

Why? because sometimes you can express the same intent using different words, why not just having standard names like `getMessages` with different options passed as arguments?

For example: `clientSubscribableMessages` mentioned in @smoya [article](https://www.asyncapi.com/blog/intent-driven-api).

I immediately ask myself:
- what "client"?
- is "subscribable" a real word? 😄 

while I could have: `getMessages`, and I could do `getMessages(operationName)`

didn't you try to fix the spec confusion around pub sub semantics with the API?

**Comment by jonaslagoni** [comment]
@derberg good question 😄 

Just to expand a bit on the article.

From the perspective of a tool developer if we do `getMessages(operationName)` you need to learn about operation pub-sub semantics and the problems that come with it. 

![temp(1)](https://user-images.githubusercontent.com/13396189/117640000-d60e6880-b184-11eb-8b60-09929a6ce8d5.png)

With the intent API we have removed the confusion offering clear functions where you do not need to know about the specification and its problems.

We have your application (application A) that needs to publish to the channel `some/channel/path`. However, if you want to describe that with AsyncAPI you need to define it from the external (what we call "client") point of view. Meaning you need to define an operation as subscribing to `some/channel/path` basically defining it for some other application B ("client" to application A). "Client" might not be the correct term to use, any suggestions? 

From the perspective of tooling, we offer the abstractions from the spec through `applicationPublishableMessages` and `clientSubscribableMessages` based on what tooling you are creating require. This works nicely with the `view` parameter some templates have adapted, since it makes it "easier" (or at least should) to understand what is happening in the code, without knowing AsyncAPI in detail.
 
![temp3](https://user-images.githubusercontent.com/13396189/117648585-af552f80-b18e-11eb-880d-05e23b6673f4.png)

If we try to map this directly to an AsyncAPI document, we can see that many of the functions access the same data. I.e. for channels that an application publishes over you have `applicationPublishableMessages` and `clientSubscribableMessages` which return the same. However, the function name has a different meaning depending on the use case. This was our way of abstracting from the problem 🙂 

Did that clarify why we suggested what we did, also in terms of your questions? 🤔 Do you see any way to improve the implicit understanding of the API? 🤔

**Reply by derberg** [reply]
The problem I see is that you did not only focus on making the API breaking-changes proof. You tried to solve the "issues" that we have with the spec through the parser API. This is something that is for me fundamentally wrong. I simply do not believe tools should fix spec, but the spec itself must be fixed. 

Example: `applicationPublishableMessages`. I assume it is a method I will use to generate my application, that AsyncAPI document describes. It will give me the list of all the messages I publish on a given channel. The descriptions of those messages, operationIds, etc will still be written from the client perspective, I will still have the spec problem there, that for example operationId name is client-focused and sounds strange in the code for the application. I hope you know what I mean 😅 

Because you focused on fixing the spec, the DX of the API was kind of neglected. "Application" and "Client" only bring additional confusion because in architecture with broker in the middle, nobody use term "client", it is just App A, B, C, etc. the app and client terms are specific and will require understanding first, before someone will be able to use the API.

> "Client" might not be the correct term to use, any suggestions?

No, not really, because as I mentioned above, I just do not believe that API should solve spec problems

> From the perspective of a tool developer if we do getMessages(operationName) you need to learn about operation pub-sub semantics and the problems that come with it.

yes, completely agree, but you need to learn about it anyway, with your API too

**Reply by jonaslagoni** [reply]
> The problem I see is that you did not only focus on making the API breaking-changes proof. You tried to solve the "issues" that we have with the spec through the parser API. This is something that is for me fundamentally wrong. I simply do not believe tools should fix spec, but the spec itself must be fixed.

@derberg we see this differently 😄 To me, we don't "fix" the specification issue, we provide a clear API that regardless of what the optimal way is to describe your AsyncAPI document is, tooling never has this issue. To me, there is no correlation, (and there shouldn't be!) between these functions and how the specification dictates the perspective. 

> No, not really, because as I mentioned above, I just do not believe that API should solve spec problems

Hmm, yea I can see how it might look like it 🤔 

Regardless of how we "fix" the specification around this issue, it would not change this API since we will always have these two perspectives from a tooling point of view. 

With version 2.0.0 of the specification, we allow users to specify how others may interact with your application. The tooling does not care what perspective the specification has, they just need to know for example the channels that the application publishes to (for code generation, message validation, etc). That's why we have two views in the API (application and client, internal and external?), cause from a tooling perspective it makes sense.

Again, I am not saying that "Application" and "Client" are the correct way to describe it, @smoya and I just had no other suggestions that made sense to use.

> The descriptions of those messages, operationIds, etc will still be written from the client perspective, I will still have the spec problem there, that for example operationId name is client-focused and sounds strange in the code for the application. I hope you know what I mean 😅

Yea, I understand 😄 To me it is a matter of good descriptions that does not imply the perspective, especially in terms of reusability.

**Reply by derberg** [reply]
there can be also other perspectives if you think about them, there can be a broker perspective too, where you use AsyncAPI to configure a broker and create topics there, and document messages floating in the system. A possible perspective that once was mentioned during SIG meeting. 

Does it mean tooling should support it? In my opinion no, just like the app perspective should not be supported as that is not expected from the spec. This is my thinking 😉 

> To me it is a matter of good descriptions

good description, just like good `operationId` is not really possible, it is either written from client perspective by following the spec or in a generic way to workaround spec challenges 🤷🏼

**Reply by jonaslagoni** [reply]
> there can be also other perspectives if you think about them, there can be a broker perspective too, where you use AsyncAPI to configure a broker and create topics there, and document messages floating in the system. A possible perspective that once was mentioned during SIG meeting.

Hmm... I see your point especially in terms of `broker perspective`, however, `application` and `client` were supposed to cover this as well, we just need some other vocabolary to describe the two perspectives. 

About configuring a broker, describing messages in systems, you would still be able to do, cause there you don't care about perspective? You are pushing the spec use-case here IMO and so you must expect tooling to not favor you 🙂 

I don't care about whether you describe it for a broker, WS (client/server), etc. For me, on the tooling side, there are only two perspectives (since the current AsyncAPI spec describes a single application). 

This means all this comes down to correctly describing `application` and `client`.

> Does it mean tooling should support it? In my opinion no, just like the app perspective should not be supported as that is not expected from the spec.

Hmmmmm... I see your point 🤔 

Just to recap, you think we should remove all the `application` intents?

**Reply by derberg** [reply]
> Just to recap, you think we should remove all the application intents?

my view on the current design is to remove any intents that have a perspective in the name.

we kind of made a round here and get back into the title for this discussion `Does intent-driven API mean the functions names need to sound like intent?`

**Reply by derberg** [reply]
My thoughts in an example:

Instead of 4 different functions: `clientPublishableChannels`, `applicationPublishableChannels()`,  `applicationSubscribableChannels()`, `clientSubscribableChannels()`

Have one function with multiple options:
- `channels()` - gives you all the channels
- `channels('sendTo')` - gives you channels that you can send to basing on perspective
- `channels('receiveFrom)` - gives you all the channels you can listen to

Where the hell is the perspective?
```js
const parser = new Parser('server')
```
or my preference
```js
parser.setPerspective('server')
```

Advantages:
- smaller and cleaner API
- you set perspective just once, less error-prone
- much easier to extend with more perspectives, as I do not see only "client"/"server" here but also "broker" that people talk about from time to time ("I just want to document all my Kafka topics in one file and not per app")
- we do not focus much on functions name, but purely on the vocabulary for perspectives and operations

I also do not see a problem if we would have multiple verbs for the same operation and perspective. As long as it is well documented and IDE friendly + makes lives of devs easier, then we should consider that.

**Reply by smoya** [reply]
@derberg 
> "Application" and "Client" only bring additional confusion because in architecture with broker in the middle, nobody use term "client", it is just App A, B, C, etc. the app and client terms are specific and will require understanding first, before someone will be able to use the API.

I agree the terminology can be improved. In fact, as per our [specification](https://www.asyncapi.com/docs/specifications/2.0.0#definitionsConsumer) we use the term `Application`, and `Producer` or `Consumer` as aggregates of the first. 
`Broker` could naturally be another aggregate. That perspective would come with few particularities in terms of business logic regarding our spec, logic that would need to be written down and handled at tooling level as well. That's the main reason for including the "perspective" (Are we using that term officially?) as part of the final intent. 

> there can be also other perspectives if you think about them, there can be a broker perspective too, where you use AsyncAPI to configure a broker and create topics there, and document messages floating in the system. A possible perspective that once was mentioned during SIG meeting.

And that was also contemplated [here](https://github.com/asyncapi/parser-js/pull/263#discussion_r608032634
). We decided that each perspective would have it's own set of intents as well (for the reason I gave in the previous quoted reply).

**Reply by smoya** [reply]
> ```js
> parser.setPerspective('server')
> ```

A quick note here: tooling users might want to use several "perspectives" at the same time so they can, for example, generate documentation or code for several "perspectives" at once. This means we would need to create several parser instances for each needed "perspective".
